src/support/serde.rs

```diff
@@ -61,54 +106,41 @@ impl<'de, const BITS: usize, const LIMBS: usize> Deserialize<'de> for Bits<BITS,
     }
 }
 
-/// Serde Visitor for human readable formats
-struct StrVisitor<const BITS: usize, const LIMBS: usize>;
+/// Serde Visitor for human readable formats.
+///
+/// Accepts either a primitive number, a decimal or a hexadecimal string.
+struct HrVisitor<const BITS: usize, const LIMBS: usize>;
 
-impl<'de, const BITS: usize, const LIMBS: usize> Visitor<'de> for StrVisitor<BITS, LIMBS> {
+impl<'de, const BITS: usize, const LIMBS: usize> Visitor<'de> for HrVisitor<BITS, LIMBS> {
     type Value = Uint<BITS, LIMBS>;
 
     fn expecting(&self, formatter: &mut Formatter) -> FmtResult {
         write!(formatter, "a {} byte hex string", nbytes(BITS))
     }
 
-    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
-    where
-        E: Error,
-    {
-        let value = trim_hex_prefix(value);
+    fn visit_u64<E: Error>(self, v: u64) -> Result<Self::Value, E> {
+        Uint::try_from(v).map_err(|_| Error::invalid_value(Unexpected::Unsigned(v), &self))
+    }
 
-        // ensure the string is the correct length (two characters in the string per
-        // byte) exception: zero, for a Uint where BITS == 0
-        if BITS == 0 {
-            // special case, this class of ints has one member, zero.
-            // zero is represented as "0x0" only
-            if value != "0" {
-                return Err(Error::invalid_value(Unexpected::Str(value), &self));
-            }
-            return Ok(Uint::<BITS, LIMBS>::ZERO);
-        }
-        if nbytes(BITS) * 2 < value.len() {
-            return Err(Error::invalid_length(value.len(), &self));
-        }
+    fn visit_u128<E: Error>(self, v: u128) -> Result<Self::Value, E> {
+        // `Unexpected::Unsigned` cannot contain a `u128`
+        Uint::try_from(v).map_err(Error::custom)
+    }
 
-        let mut limbs = [0; LIMBS];
-        for (i, chunk) in value.as_bytes().rchunks(16).enumerate() {
-            let chunk = str::from_utf8(chunk)
-                .map_err(|_| Error::invalid_value(Unexpected::Str(value), &self))?;
-            let limb = u64::from_str_radix(chunk, 16)
-                .map_err(|_| Error::invalid_value(Unexpected::Str(value), &self))?;
-            if limb == 0 {
-                continue;
-            }
-            if i >= LIMBS {
-                return Err(Error::invalid_value(Unexpected::Str(value), &self));
-            }
-            limbs[i] = limb;
+    fn visit_str<E: Error>(self, value: &str) -> Result<Self::Value, E> {
+        // Shortcut for common case
+        if value == ZERO_STR {
+            return Ok(Uint::<BITS, LIMBS>::ZERO);
         }
-        if BITS > 0 && limbs[LIMBS - 1] > Self::Value::MASK {
+        // `ZERO_STR` is the only valid serialization of `Uint<0, 0>`, so if we
```

Comment:
is this true with the more-permissive deser? do we want to use less-permissive deser for the special case?

Reply:
Yes it is true, in the two lines right below this

Reply:
what i mean is that do we still require a canonical serialization for 0-bit numbers? or do we permissively accept `0b0`, `0o0` and `0` for this type?